enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,182 @@
+2) As a `Pod`. For Openshift this option is not sufficient, so non-HA environment will run this pod under `Deployment` or as a static resource managed from the `Node` filesystem in the `openshift-cloud-controller` namespace.
+
+#### Examples
+
+##### AWS
+
+Main repository: https://github.com/kubernetes/cloud-provider-aws/
+Sample manifests: https://github.com/kubernetes/cloud-provider-aws/tree/master/manifests
+
+To try external cloud-controller-manager, siply run:
+```bash
+oc apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/master/manifests/rbac.yaml  
+oc apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/master/manifests/aws-cloud-controller-manager-daemonset.yaml
+```
+
+##### Azure
+
+Main repository: https://github.com/kubernetes-sigs/cloud-provider-azure
+Sample manifests: https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/examples/out-of-tree
+
+##### GCP
+
+Main repository: https://github.com/kubernetes/cloud-provider-gcp
+Sample manifests: https://github.com/kubernetes/cloud-provider-gcp/tree/master/deploy
+
+##### vSphere
+
+Main repository: https://github.com/kubernetes/cloud-provider-vsphere
+Sample manifests: https://github.com/kubernetes/cloud-provider-vsphere/tree/master/manifests/controller-manager
+
+### Test Plan
+
+- Each provider at the time of migration should have a working CI implemetation, which will assist in testing the provisioning with out-of-tree support enabled.
+- Ensure that transition between in-tree and out-of-tree will be handled by the `Infrastracture` `external-cloud-config` field being set.
+- Ensure in-tree and out-of-tree providers could co-exist in the cluster if this is supported.
+- Ensure that upgrade from a release running on in-tree provider only to the out-of-tree succeeds.
+
+##### Removing a deprecated feature
+
+Upstream is currently planning to remove support for in-tree providers with OCP-4.8 (kubernetes 1.21) - based on the slack [conversation](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600). The enchancement [describes](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md) steps towards this process. Our goal is to have a working alternative for in-tree providers ready to be used on-par with old implementation befor the upstream release will remove this feature from main `k/k` repository.
+
+## Timeline
+
+Follow upstream discussions and implementation/releases
+
+## Kubernetes 1.18
+
+- vSphere support graduaded to beta: https://github.com/kubernetes/enhancements/issues/670
+
+## Kubernetes 1.19
+
+- Azure support goes into beta: https://github.com/kubernetes/enhancements/issues/667
+
+## Infrastructure Needed
+
+(Possibly) Forks for out-of-tree provider repositories:
```

Comment:
Remove (Possibly), to enable us to do bug fixes in an efficient manner, we will need to have a fork of each of these projects so that we can control what we ship and port to upstream later